### PR TITLE
Implemented abscissa scale changes for LineVis

### PR DESCRIPTION
**Pull request by loichuder**
...and changed Scale toggle buttons to dropdowns.

image

### CODE REVIEW

File: src/h5web/visualizations/containers/NxSpectrumContainer.tsx [hunk 1]
```diff
@@ -35,13 +35,13 @@ function NxSpectrumContainer(props: VisContainerProps): ReactElement {
...
-  const { setScaleType } = useLineConfig();
+  const { setYScaleType } = useLineConfig();
 
   useEffect(() => {
     if (signal.scaleType) {
-      setScaleType(signal.scaleType);
+      setYScaleType(signal.scaleType);
     }
-  }, [setScaleType, signal.scaleType]);
+  }, [setYScaleType, signal.scaleType]);
```
Comment (axelboc) [inline src/h5web/visualizations/containers/NxSpectrumContainer.tsx:44]: I feel like this `useEffect` should now live in `MappedLineVis`. I know it means adding a prop, but it avoids calling `useLineConfig` in two places.

File: src/h5web/toolbar/controls/Selector.tsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  } = props;
+
+  const { height: winHeight } = useWindowSize();
+  const menuMaxHeight = winHeight - MENU_TOP - MENU_BOTTOM;
+
+  return (
+    <Wrapper className={styles.wrapper} onSelection={onChange}>
+      <Button className={styles.btn} tag="button" disabled={disabled}>
+        <div className={styles.btnLike}>
+          <span className={styles.selectedColorMap}>
```
Comment (axelboc) [inline src/h5web/toolbar/controls/Selector.tsx:64]: need generic class

File: src/h5web/toolbar/controls/Selector.tsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  options: Record<string, T[]> | T[];
+  selectedComponent: OptionComponent<T>;
+  optionComponent: OptionComponent<T>;
+}
+
+function OptionList<T extends string>(props: {
+  optionList: T[];
+  optionComponent: OptionComponent<T>;
+  value: T;
+}): ReactElement {
```
Comment (axelboc) [inline src/h5web/toolbar/controls/Selector.tsx:27]: needs an interface for the props

Comment (axelboc) [reply src/h5web/toolbar/controls/Selector.tsx:27]: Might be more readable to have this child component live in a separate file.

File: src/h5web/toolbar/controls/ColorMapSelector.tsx [hunk 4]
```diff
@@ -1,31 +1,30 @@
...
-  interpolator: D3Interpolator;
+interface OptionProps {
+  option: ColorMap;
 }
 
-function Gradient(props: GradientProps): ReactElement {
-  const { interpolator } = props;
-  const backgroundImage = generateCSSLinearGradient(interpolator, 'right');
-  return <div className={styles.gradient} style={{ backgroundImage }} />;
+function ColormapOption(props: OptionProps): ReactElement {
```
Comment (axelboc) [inline src/h5web/toolbar/controls/ColorMapSelector.tsx:16]: `ColorMapOption`

Comment (axelboc) [reply src/h5web/toolbar/controls/ColorMapSelector.tsx:16]: you could inline `OptionProps` like you did in the scale type selector

File: src/h5web/toolbar/controls/ColorMapSelector.tsx [hunk 5]
```diff
@@ -37,53 +36,15 @@ interface Props {
...
-        </ul>
-      </Menu>
-    </Wrapper>
+    <Selector
+      value={value}
+      disabled={disabled}
+      onChange={onChange}
+      options={COLORMAP_GROUPS}
+      selectedComponent={ColormapOption}
+      optionComponent={ColormapOption}
```
Comment (axelboc) [inline src/h5web/toolbar/controls/ColorMapSelector.tsx:46]: It seems that `selectedComponent` and `optionComponent` are always the same, in the end. Is it worth keeping them as separate props?

File: src/h5web/visualizations/heatmap/interpolators.ts [hunk 6]
```diff
@@ -102,10 +103,10 @@ export const INTERPOLATORS = {
...
-  'Single hue': SINGLE_HUE,
-  'Multi hue': MULTI_HUE,
-  Cyclical: CYCLICAL,
-  Diverging: DIVERGING,
+export const COLORMAP_GROUPS = {
+  Common: Object.keys(MOST_USED) as ColorMap[],
+  'Single hue': Object.keys(SINGLE_HUE) as ColorMap[],
+  'Multi hue': Object.keys(MULTI_HUE) as ColorMap[],
+  Cyclical: Object.keys(CYCLICAL) as ColorMap[],
+  Diverging: Object.keys(DIVERGING) as ColorMap[],
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/interpolators.ts:111]: Hmm you must be getting a cyclic dependency error from using `ColorMap` here, aren't you??

Comment (axelboc) [reply src/h5web/visualizations/heatmap/interpolators.ts:111]: You should be able to get around it by casting `option` to `ColorMap` in `ColorMapOption`.

**Review by axelboc — APPROVED** [review]
Looks great!!

**Review by axelboc — APPROVED** [review]
Very nice! The new test looks great, and the two `useEffect` hooks look fine in `MappedLineVis`.

Only thing I see would be to stop persisting the scale types as we discussed.

**Comment by loichuder** [comment]
> Only thing I see would be to stop persisting the scale types as we discussed.

**Pull request merged**